### PR TITLE
Apply a suggestion cleanly when there are references

### PR DESCRIPTION
**Pull request by paulfitz**
Accepting a proposal used to write the diff out a piece at a time, one engine call per row or cell. A failure partway left the earlier writes in place and marked the proposal applied anyway, so Accept vanished with the job half done. Separately, a reference to a row the proposal itself added carried a row id from the other document, and landed on an unrelated row.

Patch now plans every action first and sends them in one call, giving added rows negative temporary ids for the engine to resolve. Those resolve as rows land, so a reference may only name a row added earlier; rather than order the tables to suit, which is impossible for a cycle, we hold such references out of their add and set them by a later update.

Also: refuse truncated summaries, whose dropped cells would land as nulls; skip updates to rows the proposal also removes; count a proposal applied only when all of it applied; give the patch log a small polish; correct an engine message that called an unresolved temporary id unresolvable when the row simply had not been added yet; and import `reportError` on the proposals page, where it had been resolving to the DOM global of the same name, so apply failures showed the user nothing at all.

### CODE REVIEW

**Comment by paulfitz** [comment]
(Now that #2385 has landed, the row id recycling issue is cleaned up - updated test)

File: test/nbrowser/ProposedChangesPage.ts [hunk 1]
```diff
@@ -801,6 +801,113 @@ describe("ProposedChangesPage", function() {
     await returnToTrunk(url);
   });
 
+  // The sibling tests above accept while the trunk stands still, so both
```
Comment (berhalak) [inline test/nbrowser/ProposedChangesPage.ts:804]: Can you rephrase `The sibling tests above accept while ...`?

File: test/nbrowser/ProposedChangesPage.ts [hunk 2]
```diff
@@ -801,6 +801,113 @@ describe("ProposedChangesPage", function() {
...
+
+    await returnToTrunk(url);
+  });
+
+  it("leaves the trunk alone when a proposal no longer applies", async function() {
+    const { api, doc } = await makeLifeDoc();
+    const url = await driver.getCurrentUrl();
+
+    await workOnCopy(url);
+    // Add a row and edit another. The add is the one to watch: nothing is
```
Comment (berhalak) [inline test/nbrowser/ProposedChangesPage.ts:869]: Can you also rephrase this one?

File: test/nbrowser/ProposedChangesPage.ts [hunk 3]
```diff
@@ -801,6 +801,113 @@ describe("ProposedChangesPage", function() {
...
+    // wrong with it, and it must still not land.
+    await gu.getCell("B", 3).click();
```
Comment (berhalak) [inline test/nbrowser/ProposedChangesPage.ts:871]: It would be easier to not use UI to write those rows, regular actions would be more readable.

File: app/server/lib/Patch.ts [hunk 4]
```diff
@@ -1,198 +1,220 @@
...
+
+    const actions: UserAction[] = [];
+    const heldBack: RowCells[] = [];
+    for (const group of groupByColumnSet(additions, ra => ra.cols)) {
+      const groupCols = group[0].cols;
+      const colValues: Record<string, CellValue[]> = {};
+      for (const colId of groupCols) { colValues[colId] = []; }
+      for (const ra of group) {
+        const cells: Record<string, CellValue> = {};
+        for (const colId of groupCols) {
```
Comment (berhalak) [inline app/server/lib/Patch.ts:174]: There are a lot of for(s) here, any chance that this and other methods here can use some functional style? I also haven't found test cases for all those methods. Most of methods here  could be static methods in module level with unit tests. It is easier to review the test case.

File: app/server/lib/Patch.ts [hunk 5]
```diff
@@ -1,198 +1,220 @@
...
-        if (!cellDelta) {
-          changes.push({
-            msg: "there is a row that does not exist anymore",
-          });
-          continue;
+    const colIds = Object.keys(columnDeltas).filter(colId => !this._shouldSkip(tableId, colId));
+
+    const additions: RowAdd[] = rows.map(row => ({
+      sourceRowId: row,
+      // Negating the source id keeps temp ids distinct within the table,
```
Comment (berhalak) [inline app/server/lib/Patch.ts]: The single comment in this method is attached to its least complicated part :)

File: app/server/lib/Patch.ts [hunk 6]
```diff
@@ -1,198 +1,220 @@
...
+          const translated = translateRefValue(this._getRefInfo(tableId, colId), value, isAddedRow);
+          if (!translated) {
+            colValues[colId].push(value);
+          } else {
+            cells[colId] = translated.value;
+            colValues[colId].push(null);
+          }
         }
-        rec[colId] = cellDelta[0]?.[0];
+        if (Object.keys(cells).length > 0) { heldBack.push({ rowId: ra.tempRowId, cells }); }
```
Comment (berhalak) [inline app/server/lib/Patch.ts:184]: Why aren't we storing update actions in the heldBack collection?

File: app/server/lib/Patch.ts (hunk 4 above)
Comment (berhalak) [reply app/server/lib/Patch.ts:174]: I think the `cells` variable name (and `ra`) is what prevents me from understending this method. I'd vote to change the `cells` to something more meaningful that explains why we don't update it in the first branch of the condition below.

File: app/server/lib/Patch.ts [hunk 7]
```diff
@@ -230,4 +252,58 @@ export class Patch {
     this._columnsByTableIdAndColId[tableId] = Object.fromEntries(columns.map(rec => [String(rec.colId), rec]));
     return this._columnsByTableIdAndColId[tableId];
   }
+
+  private _getRefInfo(tableId: string, colId: string): RefInfo {
+    const inner = getSetMapValue(this._refInfoByTableAndCol, tableId, () => new Map<string, RefInfo>());
```
Comment (berhalak) [inline app/server/lib/Patch.ts:257]: Are you sure that this cache is needed? It looks like a cache of a `_columnsByTableIdAndColId` cache (so double cache).

File: app/server/lib/RefTranslator.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+ * resolve. Rows it does not add are taken to be shared with the common
+ * ancestor, so their ids mean the same thing on both sides and pass
+ * through untouched.
+ */
+
+import { CellValue } from "app/common/DocActions";
+import { extractInfoFromColType, isList } from "app/common/gristTypes";
+import { GristObjCode } from "app/plugin/GristData";
+
+export type RefInfo =
```
Comment (berhalak) [inline app/server/lib/RefTranslator.ts:16]: Nitpick: It looks like unnecessary abstraction - it is almost identical to null | "Ref:xxx" | "RefList:xxx".

File: sandbox/grist/test_temp_rowids.py [hunk 9]
```diff
@@ -206,18 +206,60 @@ def test_reflist_rejects_unresolved_temp_id(self):
...
   def test_ref_rejects_unresolved_temp_id(self):
     self.load_sample(testsamples.sample_students)
 
     # Same rejection for a single Ref, not just RefList.
-    with self.assertRaisesRegex(ValueError, r"unknown temporary row id -99 in column Schools\.address"):
+    with self.assertRaisesRegex(ValueError, r"temporary row id -99 in column Schools\.address"):
       self.engine.apply_user_actions([useractions.from_repr(
         ['AddRecord', 'Schools', None, {'name': 'S1', 'address': -99}])])
 
+  def test_refs_must_point_backwards(self):
```
Comment (berhalak) [inline sandbox/grist/test_temp_rowids.py:221]: Why are you adding a test when no python code was changed?

**Review by berhalak — CHANGES_REQUESTED** [review]
Couldn't find anything - apart from edge cases with two-way references (that were also before). Only comments about structure/comments.

File: sandbox/grist/test_temp_rowids.py (hunk 9 above)
Comment (paulfitz) [reply sandbox/grist/test_temp_rowids.py]: Pins down an ordering rule we depend on. Added comment.
